parser/Parser.idr

```diff
@@ -4,11 +4,129 @@ import Backend
 import Backend.Utils
 import Backend.Haskell
 
+import Options.Applicative
+
+%default total
+
+flag' : a -> (Option FlagParams a -> Option FlagParams a) -> Parser a
+flag' d f = OptP (f $ Opt defProps (FlagReader [] d))
+
+
+data InputFile = StringInput String
+               | FileInput String
+
+Show InputFile where
+  show StringInput = "command line input"
+  show (FileInput str) = str
+
+data OutputFile = StdOutput
+                | FileOutput String
+
+Show OutputFile where
+  show StdOutput = "stdout"
+  show (FileOutput str) = str
+
+stringInput : Parser InputFile
+stringInput = StringInput <$> strOption (long "inline" . short 'i')
+
+fileInput : Parser InputFile
+fileInput = FileInput <$> strOption (long "source" . short 's')
+
+parseInput : Parser InputFile
+parseInput = stringInput <|> fileInput
+
+stdOutput : Parser OutputFile
+stdOutput = flag' StdOutput (long "stdout")
+
+fileOutput : Parser OutputFile
+fileOutput = FileOutput <$> strOption (long "dest" . short 'd')
+
+parseOutput : Parser OutputFile
+parseOutput = fileOutput <|> stdOutput <|> pure StdOutput
+
+
+record TypedefOpts where
+  constructor MkTypedefOpts
+  input : InputFile
+  output : OutputFile
+
+data CommandLineOpts = Help | GenerateTDefOpt TypedefOpts | HelpFallback
+
+Show TypedefOpts where
+  show (MkTypedefOpts i o) = "input : " ++ show i ++ "output : " ++ show o
+
+helpMessage : String
+helpMessage = """
+Welcome to Typedefs, programming language for types.
+
+Usage:
+  typedefs_parser (-i INLINE_TDEF | -s SOURCE) [-d DEST | --stdout]
+
+  --source, -s : path to the input file
+    example:
+      typedefs_parser -s bool.tdef
+
+  --inline, -i : input as a string inline
+    example:
+      typedefs_parser -i "(name bool (+ 1 1))" --stdout
+
+  --dest, -d : path to the destination file
+    example:
+      typedefs_parser -i bool.tdef project/typedefs/types.hs
```

Comment:
Shouldn't this be `-s` instead of `-i`? Same for the example below. Also `-d` seems to be missing.

Reply:
You're right, it should either be 
`-i "(+ 1 1)"` or `-s bool.tdef`

Reply:
should be fixed